File: cs/src/DataCentric/Platform/DataSource/DataSourceData.cs
/*
Copyright (C) 2013-present The DataCentric Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.Collections.Generic;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using NodaTime;

namespace DataCentric
{
    /// <summary>
    /// Data source is a logical concept similar to database
    /// that can be implemented for a document DB, relational DB,
    /// key-value store, or filesystem.
    ///
    /// Data source API provides the ability to:
    ///
    /// (a) store and query datasets;
    /// (b) store records in a specific dataset; and
    /// (c) query record across a group of datasets.
    ///
    /// This record is stored in root dataset.
    /// </summary>
    public abstract class DataSourceData : RootRecord<DataSourceKey, DataSourceData>, IDataSource
    {
        //--- FIELDS

        /// <summary>
        /// Dictionary of dataset ObjectIds stored under string dataSetName.
        /// </summary>
        private Dictionary<string, ObjectId> dataSetDict_ { get; } = new Dictionary<string, ObjectId>();

        /// <summary>
        /// Dictionary of the expanded list of import ObjectIds for each dataset ObjectId,
        /// including imports of imports to unlimited depth with cyclic references and
        /// duplicates removed.
        /// </summary>
        private Dictionary<ObjectId, HashSet<ObjectId>> importDict_ { get; } = new Dictionary<ObjectId, HashSet<ObjectId>>();

        //--- ELEMENTS

        /// <summary>Unique data source name.</summary>
        [BsonRequired]
        public string DataSourceName { get; set; }

        /// <summary>
        /// This class enforces strict naming conventions
        /// for database naming. While format of the resulting database
        /// name is specific to data store type, it always consists
        /// of three tokens: InstanceType, InstanceName, and EnvName.
        /// The meaning of InstanceName and EnvName tokens depends on
        /// the value of InstanceType enumeration.
        /// </summary>
        public DbNameKey DbName { get; set; }

        /// <summary>
        /// Identifies the data store used by this data source.
        ///
        /// Data store represents a database server or similar concept for non-database
        /// storage. It is not the same as data source (database) as it only specifies
        /// the server, and each server can host multiple data sources (databases).
        ///
        /// Separating the data store from the data source helps separate server
        /// specifics such as URI, connection string, etc in data store from the
        /// definition of how the data is stored on the server, including the
        /// environment (which usually maps to database name) and data representation
        /// (basic or temporal).
        /// </summary>
        public DataStoreKey DataStore { get; set; }

        /// <summary>
        /// Use this flag to mark dataset as readonly, but use either
        /// IsReadOnly() or CheckNotReadonly() method to determine the
        /// readonly status because the dataset may be readonly for
        /// two reasons:
        ///
        /// * ReadOnly flag is true; or
        /// * One of SavedByTime or SavedById is set
        /// </summary>
        public bool ReadOnly { get; set; }

        //--- ABSTRACT

        /// <summary>
        /// Returns true if the data source is readonly,
        /// which may be because ReadOnly flag is true,
        /// or due to other flags (e.g. SavedBy) defined
        /// in derived types.
        /// </summary>
        public abstract bool IsReadOnly();

        /// <summary>
        /// The returned ObjectIds have the following order guarantees:
        ///
        /// * For this data source instance, to arbitrary resolution; and
        /// * Across all processes and machines, to one second resolution
        ///
        /// One second resolution means that two ObjectIds created within
        /// the same second by different instances of the data source
        /// class may not be ordered chronologically unless they are at
        /// least one second apart.
        /// </summary>
        public abstract ObjectId CreateOrderedObjectId();

        /// <summary>
        /// Load record by its ObjectId.
        ///
        /// Return null if there is no record for the specified ObjectId;
        /// however an exception will be thrown if the record exists but
        /// is not derived from TRecord.
        /// </summary>
        public abstract TRecord LoadOrNull<TRecord>(ObjectId id)
            where TRecord : Record;

        /// <summary>
        /// Load record by string key from the specified dataset or
        /// its list of imports. The lookup occurs first in descending
        /// order of dataset ObjectIds, and then in the descending
        /// order of record ObjectIds within the first dataset that
        /// has at least one record. Both dataset and record ObjectIds
        /// are ordered chronologically to one second resolution,
        /// and are unique within the database server or cluster.
        ///
        /// The root dataset has empty ObjectId value that is less
        /// than any other ObjectId value. Accordingly, the root
        /// dataset is the last one in the lookup order of datasets.
        ///
        /// The first record in this lookup order is returned, or null
        /// if no records are found or if DeletedRecord is the first
        /// record.
        ///
        /// Return null if there is no record for the specified ObjectId;
        /// however an exception will be thrown if the record exists but
        /// is not derived from TRecord.
        /// </summary>
        public abstract TRecord LoadOrNull<TKey, TRecord>(TypedKey<TKey, TRecord> key, ObjectId loadFrom)
            where TKey : TypedKey<TKey, TRecord>, new()
            where TRecord : TypedRecord<TKey, TRecord>;

        /// <summary>
        /// Get query for the specified type.
        ///
        /// After applying query parameters, the lookup occurs first in
        /// descending order of dataset ObjectIds, and then in the descending
        /// order of record ObjectIds within the first dataset that
        /// has at least one record. Both dataset and record ObjectIds
        /// are ordered chronologically to one second resolution,
        /// and are unique within the database server or cluster.
        ///
        /// The root dataset has empty ObjectId value that is less
        /// than any other ObjectId value. Accordingly, the root
        /// dataset is the last one in the lookup order of datasets.
        ///
        /// Generic parameter TRecord is not necessarily the root data type;
        /// it may also be a type derived from the root data type.
        /// </summary>
        public abstract IQuery<TRecord> GetQuery<TRecord>(ObjectId loadFrom)
            where TRecord : Record;

        /// <summary>
        /// Save record to the specified dataset. After the method exits,
        /// record.DataSet will be set to the value of the dataSet parameter.
        ///
        /// All Save methods ignore the value of record.DataSet before the
        /// Save method is called. When dataset is not specified explicitly,
        /// the value of dataset from the context, not from the record, is used.
        /// The reason for this behavior is that the record may be stored from
        /// a different dataset than the one where it is used.
        ///
        /// This method guarantees that ObjectIds will be in strictly increasing
        /// order for this instance of the data source class always, and across
        /// all processes and machine if they are not created within the same
        /// second.
        /// </summary>
        public abstract void Save<TRecord>(TRecord record, ObjectId saveTo)
            where TRecord : Record;

        /// <summary>
        /// Write a DeletedRecord in deleteIn dataset for the specified key
        /// instead of actually deleting the record. This ensures that
        /// a record in another dataset does not become visible during
        /// lookup in a sequence of datasets.
        ///
        /// To avoid an additional roundtrip to the data store, the delete
        /// marker is written even when the record does not exist.
        /// </summary>
        public abstract void Delete<TKey, TRecord>(TypedKey<TKey, TRecord> key, ObjectId deleteIn)
            where TKey : TypedKey<TKey, TRecord>, new()
            where TRecord : TypedRecord<TKey, TRecord>;

        /// <summary>
        /// Permanently deletes (drops) the database with all records
        /// in it without the possibility to recover them later.
        ///
        /// This method should only be used to free storage. For
        /// all other purposes, methods that preserve history should
        /// be used.
        ///
        /// ATTENTION - THIS METHOD WILL DELETE ALL DATA WITHOUT
        /// THE POSSIBILITY OF RECOVERY. USE WITH CAUTION.
        /// </summary>
        public abstract void DeleteDb();

        //--- METHODS

        /// <summary>
        /// Error message if the data source is readonly,
        /// which may be for the following reasons:
        ///
        /// * ReadOnly flag is true; or
        /// * One of SavedByTime or SavedById is set
        /// </summary>
        public void CheckNotReadOnly()
        {
            if (IsReadOnly())
                throw new Exception(
                    $"Attempting write operation for readonly data source {DataSourceName}. " +
                    $"A data source is readonly if either (a) its ReadOnly flag is set, or (b) " +
                    $"one of SavedByTime or SavedById is set.");
        }

        /// <summary>
        /// Get ObjectId of the dataset with the specified name.
        ///
        /// All of the previously requested dataSetIds are cached by
        /// the data source. To load the latest version of the dataset
        /// written by a separate process, clear the cache first by
        /// calling DataSource.ClearDataSetCache() method.
        ///
        /// Returns null if not found.
        /// </summary>
        public ObjectId? GetDataSetOrNull(string dataSetName, ObjectId loadFrom)
        {
            if (dataSetDict_.TryGetValue(dataSetName, out ObjectId result))
            {
                // Check if already cached, return if found
                return result;
            }
            else
            {
                // Otherwise load from storage (this also updates the dictionaries)
                DataSetKey dataSetKey = new DataSetKey() { DataSetName = dataSetName };
                DataSetData dataSetData = this.LoadOrNull(dataSetKey, loadFrom);

                // If not found, return ObjectId.Empty
                if (dataSetData == null) return null;

                // If found, cache result in ObjectId dictionary
                dataSetDict_[dataSetName] = dataSetData.Id;

                // Build and cache dataset lookup list if not found
                if (!importDict_.TryGetValue(dataSetData.Id, out HashSet<ObjectId> importSet))
                {
                    importSet = BuildDataSetLookupList(dataSetData);
                    importDict_.Add(dataSetData.Id, importSet);
                }

                return dataSetData.Id;
            }
        }

        /// <summary>
        /// Save new version of the dataset.
        ///
        /// This method sets Id element of the argument to be the
        /// new ObjectId assigned to the record when it is saved.
        /// The timestamp of the new ObjectId is the current time.
        ///
        /// This method updates in-memory cache to the saved dataset.
        /// </summary>
        public void SaveDataSet(DataSetData dataSetData, ObjectId saveTo)
        {
            // Save dataset to storage. This updates its Id
            // to the new ObjectId created during save
            Save<DataSetData>(dataSetData, saveTo);

            // Update dataset dictionary with the new Id
            dataSetDict_[dataSetData.Key] = dataSetData.Id;

            // Update lookup list dictionary
            var lookupList = BuildDataSetLookupList(dataSetData);
            importDict_.Add(dataSetData.Id, lookupList);
        }

        /// <summary>
        /// Returns enumeration of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed.
        ///
        /// The list will not include datasets that are after the value of
        /// SavedByTime/SavedById if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        /// </summary>
        public IEnumerable<ObjectId> GetDataSetLookupList(ObjectId loadFrom)
        {
            // Root dataset has no imports (there is not even a record
            // where these imports can be specified).
            //
            // Return list containing only the root dataset (ObjectId.Empty) and exit
            if (loadFrom == ObjectId.Empty)
            {
                return new ObjectId[] {ObjectId.Empty};
            }

            if (importDict_.TryGetValue(loadFrom, out HashSet<ObjectId> result))
            {
                // Check if the lookup list is already cached, return if yes
                return result;
            }
            else
            {
                // Otherwise load from storage (returns null if not found)
                DataSetData dataSetData = LoadOrNull<DataSetData>(loadFrom);

                if (dataSetData == null) throw new Exception($"Dataset with ObjectId={loadFrom} is not found.");
                if (dataSetData.DataSet != ObjectId.Empty) throw new Exception($"Dataset with ObjectId={loadFrom} is not stored in root dataset.");

                // Build the lookup list
                result = BuildDataSetLookupList(dataSetData);

                // Add to dictionary and return
                importDict_.Add(loadFrom, result);
                return result;
            }
        }

        //--- PROTECTED

        /// <summary>
        /// Records where _id is greater than the returned value will be
        /// ignored by the data source.
        ///
        /// This element is set based on either SavedByTime and SavedById
        /// elements that are alternates; only one of them can be specified.
        /// </summary>
        protected abstract ObjectId? GetSavedBy();

        //--- PRIVATE

        /// <summary>
        /// Builds hashset of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed. This method uses cached lookup
        /// list for the import datasets but not for the argument dataset.
        ///
        /// The list will not include datasets that are after the value of
        /// SavedByTime/SavedById if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        ///
        /// This overload of the method will return the result hashset.
        ///
        /// This private helper method should not be used directly.
        /// It provides functionality for the public API of this class.
        /// </summary>
        private HashSet<ObjectId> BuildDataSetLookupList(DataSetData dataSetData)
        {
            // Delegate to the second overload
            var result = new HashSet<ObjectId>();
            BuildDataSetLookupList(dataSetData, result);
            return result;
        }

        /// <summary>
        /// Builds hashset of import datasets for specified dataset data,
        /// including imports of imports to unlimited depth with cyclic
        /// references and duplicates removed. This method uses cached lookup
        /// list for the import datasets but not for the argument dataset.
        ///
        /// The list will not include datasets that are after the value of
        /// SavedByTime/SavedById if specified, or their imports (including
        /// even those imports that are earlier than the constraint).
        ///
        /// This overload of the method will return the result hashset.
        ///
        /// This private helper method should not be used directly.
        /// It provides functionality for the public API of this class.
        /// </summary>
        private void BuildDataSetLookupList(DataSetData dataSetData, HashSet<ObjectId> result)
        {
            // Return if the dataset is null or has no imports
            if (dataSetData == null) return;

            // Error message if dataset has no Id or Key set
            dataSetData.Id.CheckHasValue();
            dataSetData.Key.CheckHasValue();

            var savedBy = GetSavedBy();
            if (savedBy != null && dataSetData.Id > savedBy.Value)
            {
                // Do not add if revision time constraint is set and is before this dataset.
                // In this case the import datasets should not be added either, even if they
                // do not fail the revision time constraint
                return;
            }

            // Add self to the result
            result.Add(dataSetData.Id);

            // Add imports to the result
            if (dataSetData.Imports != null)
            {
                foreach (var dataSetId in dataSetData.Imports)
                {
                    // Dataset cannot include itself as its import
                    if (dataSetData.Id == dataSetId)
                        throw new Exception(
                            $"Dataset {dataSetData.Key} with ObjectId={dataSetData.Id} includes itself in the list of its imports.");

                    // The Add method returns true if the argument is not yet present in the hashset
                    if (result.Add(dataSetId))
                    {
                        // Add recursively if not already present in the hashset
                        var cachedImportList = GetDataSetLookupList(dataSetId);
                        foreach (var importId in cachedImportList)
                        {
                            result.Add(importId);
                        }
                    }
                }
            }
        }
    }
}

File: py/src/datacentric/platform/data_source/data_source_data.py
from abc import ABC, abstractmethod

from bson.objectid import ObjectId
from typing import List, Set, Dict, Iterable, Union
from datacentric.platform.data_set import DataSetData
from datacentric.platform.data_set.data_set import DataSetKey
from datacentric.types.record import RootRecord, Record, Key


class DataSourceData(RootRecord, ABC):
    empty_id = ObjectId('000000000000000000000000')
    common_id = 'Common'

    def __init__(self):
        RootRecord.__init__(self)
        self._data_set_dict = dict()  # type: Dict[str, ObjectId]
        self._import_dict = dict()  # type: Dict[ObjectId, Set[ObjectId]]

        self.data_source_id = None  # type: str

        self.db_name = None  # type: str
        self.data_store = None  # type: str
        self.readonly = None  # type: bool

    @abstractmethod
    def create_ordered_object_id(self) -> ObjectId:
        pass

    @abstractmethod
    def is_readonly(self) -> bool:
        pass

    def check_not_readonly(self):
        if self.is_readonly():
            raise Exception(f'Attempting write operation for readonly data source {self.data_source_id}. '
                            f'A data source is readonly if either (a) its ReadOnly flag is set, or (b) '
                            f'one of SavedByTime or SavedById is set.')

    @abstractmethod
    def load_or_null(self, id_: ObjectId) -> Record:
        pass

    @abstractmethod
    def reload_or_null(self, key: Key, load_from: ObjectId) -> Record:
        pass

    @abstractmethod
    def get_query(self, load_from: ObjectId):
        pass

    @abstractmethod
    def save(self, record: Record, save_to: ObjectId) -> None:
        pass

    @abstractmethod
    def delete(self, key: Key, delete_in: ObjectId) -> None:
        pass

    @abstractmethod
    def delete_db(self) -> None:
        pass

    def get_data_set_or_empty(self, data_set_id: str, load_from: ObjectId) -> ObjectId:
        if data_set_id in self._data_set_dict:
            return self._data_set_dict[data_set_id]
        else:
            return self.__load_data_set_or_empty(data_set_id, load_from)

    def save_data_set(self, data_set_data: DataSetData, save_to: ObjectId) -> None:
        self.save(data_set_data, save_to)
        self._data_set_dict[data_set_data.key_] = data_set_data.id_
        lookup_list = self.__build_data_set_lookup_list(data_set_data)
        self._import_dict[data_set_data.id_] = lookup_list

    def get_data_set_lookup_list(self, load_from: ObjectId) -> Iterable[ObjectId]:
        if load_from == self.empty_id:
            return [self.empty_id]
        if load_from in self._import_dict:
            return self._import_dict[load_from]
        else:
            data_set_data = self.load_or_null(load_from)  # type: DataSetData
            if data_set_data is None:
                raise Exception(f'Dataset with ObjectId={load_from} is not found.')
            if data_set_data.data_set != self.empty_id:
                raise Exception(f'Dataset with ObjectId={load_from} is not stored in root dataset.')
            result = self.__build_data_set_lookup_list(data_set_data)
            self._import_dict[load_from] = result
            return result

    @abstractmethod
    def _get_saved_by(self) -> ObjectId:
        pass

    def __load_data_set_or_empty(self, data_set_id: str, load_from: ObjectId) -> ObjectId:
        data_set_key = DataSetKey(data_set_id)
        data_set_data = self.load_or_null_by_key(data_set_key, load_from)  # type: DataSetData

        if data_set_data is None:
            return self.empty_id
        self._data_set_dict[data_set_id] = data_set_data.id_
        if data_set_data.id_ not in self._import_dict:
            import_set = self.__build_data_set_lookup_list(data_set_data)
            self._import_dict[data_set_data.id_] = import_set
        return data_set_data.id_

    def __build_data_set_lookup_list(self, data_set_data: DataSetData) -> Union[Set[ObjectId], None]:
        if data_set_data is None:
            return

        if not ObjectId.is_valid(data_set_data.id_):
            raise Exception('Required ObjectId value is not set.')
        if data_set_data.key_ is None or data_set_data.key_ == '':
            raise Exception('Required string value is not set.')

        saved_by = self._get_saved_by()
        if saved_by is not None and data_set_data.id_ > saved_by:
            return

        result = set()
        result.add(data_set_data.id_)
        if data_set_data.imports is not None:
            for data_set_id in data_set_data.imports:
                if data_set_data.id_ == data_set_id:
                    raise Exception(f'Dataset {data_set_data.key_} with ObjectId={data_set_data.id_} '
                                    f'includes itself in the list of its imports.')
                if data_set_id not in result:
                    result.add(data_set_id)
                    cached_import_list = self.get_data_set_lookup_list(data_set_id)
                    for import_id in cached_import_list:
                        result.add(import_id)
                else:
                    result.add(data_set_id)

    # From extensions
    def load(self, id_: ObjectId) -> Record:
        result = self.load_or_null(id_)
        if result is None:
            raise Exception(f'Record with ObjectId={id_} is not found in data store {self.data_source_id}.')
        return result

    # renamed
    def load_by_key(self, key_: Key, load_from: ObjectId):
        return key_.load(self.context, load_from)

    # renamed
    def load_or_null_by_key(self, key_: Key, load_from: ObjectId):
        return key_.load_or_null(self.context, load_from)

    def get_data_set(self, data_set_id: str, load_from: ObjectId):
        result = self.get_data_set_or_empty(data_set_id, load_from)
        if result is None:
            raise Exception(f'Dataset {data_set_id} is not found in data store {self.data_source_id}.')

    def get_common(self):
        return self.get_data_set(self.common_id, self.empty_id)

    def create_common(self) -> ObjectId:
        result = DataSetData()
        result.data_set_id = self.common_id
        self.save_data_set(result, self.empty_id)
        return result.id_

    def create_data_set(self, data_set_id: str, save_to: ObjectId, import_data_sets: List[ObjectId]) -> ObjectId:
        result = DataSetData()
        result.data_set_id = data_set_id

        if import_data_sets is not None:
            result.imports = [x for x in import_data_sets]

        self.save_data_set(result, save_to)
        return result.id_
